### DISCUSSION TITLE
v2.0.0-alpha-1 🌈

### DISCUSSION THREAD

**Discussion by bitwiseman** [hub4j/github-api#1938] [General]
## Major version bump to v2.x

I would love for github-api v2.x to be a larger change that has major new features.

However, the most pressing need is clearing technical debt for better stability and ease of adding new features. These changes break binary compatibility requiring a major version bump.

v2.0 includes:

* Drop support for Java 8
* Remove functional dependencies on HttpURLConnection
* Remove most deprecated classes and methods
* Remove all "bridge methods"

## Changes

- Prepare release (bitwiseman): github-api-2.0.0-alpha-1 @bitwiseman (#1936)
- Begin 2.x release train @bitwiseman (#1935)
- Prepare release (bitwiseman): github-api-1.325 @github-actions (#1932)

<hr /><em>This discussion was created from the release <a href='https://github.com/hub4j/github-api/releases/tag/github-api-2.0.0-alpha-1'>v2.0.0-alpha-1 🌈</a>.</em>

**Comment by bitwiseman** [comment]
@gsmet @ihrigb @mestebangutierrez @tginiotis-at-work @kgromov @MarkEWaite @alecharp 
Any feedback you can provide would be appreciated.

**Reply by ihrigb** [reply]
I finally got to test the alpha and I am happy with it. Only thing I stumbled accross is that the method `GHRepository.getPullRequests(GHIssueState)` has been removed and has not been marked as deprecated before. Changing to `queryPullRequests` is easy though.

**Reply by gsmet** [reply]
@bitwiseman yeah I need to have a look and try to port the Quarkus extension and Quarkus GitHub App to this new version. It's on my TODO list for soonish.

**Reply by bitwiseman** [reply]
@ihrigb 
Filed as #1957

**Reply by gsmet** [reply]
@bitwiseman I was finally able to do some testing. From my perspective, it's all good: I ported Quarkus GitHub API and Quarkus GitHub App to it with only very minor changes and tested our GitHub bot both in JVM mode and native.

Only one thing comes to mind: it seems you dropped the unbridged artifact? I think you should reinstate it even if for now you don't have bridges as you will add some at some point and they are going to be problematic again for Mockito users.

**Reply by gsmet** [reply]
Oh and also, it seems you didn't drop all the `IOException` that are extremely annoying when coding with the API. It has been one of my major complaint over time, as it's a bit painful to have to throw `IOException` everywhere.

In some cases, we even have `IOException` for methods that are definitely not throwing anything such as:

https://github.com/hub4j/github-api/blob/71501216c7f15f21bc598c6ffac94339e6ad5057/src/main/java/org/kohsuke/github/GHDiscussion.java#L42-L44

I think we have a unique opportunity here to clean this up.

**Reply by gsmet** [reply]
BTW, I'm aware it's going to require app changes but really it's for the better and given we are doing a major, I think it's probably now or never. People who don't want to handle this in their apps can keep using the old version if they prefer.

**Reply by bitwiseman** [reply]
> Only one thing comes to mind: it seems you dropped the unbridged artifact? 

There was a glitch that I noticed a few weeks after standard release of 1.325.  I believe I fixed it with https://github.com/hub4j/github-api/commit/bfa107f4d9569d61edfe2d30e8d61cdf7db7190a and https://github.com/hub4j/github-api/commit/19ba058ee4c7f078ad0403715d6c6f867906be0a .
Maven Central has 1.325 and 1.326 under https://mvnrepository.com/artifact/org.kohsuke/github-api-unbridged . 

> Oh and also, it seems you didn't drop all the IOException that are extremely annoying when coding with the API. 

This was intentional.  As note it is disappointing, but I'm trying to walk the line for minimal number of binary breaking changes for the sake of Jenkins usage.  Jenkins' plugin behavior means that compile-time breaking changes are less problematic than binary breaking changes - thus usage of bridge methods.  Looking at [japicmp - METHOD_NO_LONGER_THROWS_CHECKED_EXCEPTION](https://siom79.github.io/japicmp/MavenPlugin.html) it appears that removing checked exceptions is a compile-time breaking change but not a binary breaking change.  If verified, then I'd be very comfortable removing most checked exceptions as part 2.0.

**Reply by gsmet** [reply]
> If verified, then I'd be very comfortable removing most checked exceptions as part 2.0.

I think it would be worth checking. We had quite a lot of complaints about that over time and every time I demo the API at a conference, it does look unnecessarily verbose (which might seem like a small price to pay but we are all struggling to make our API more user friendly and it's definitely worth it).
It's also sometimes quite annoying when dealing with the Stream API and having to catch the exceptions in the lambda. I got hit by that several times and see https://github.com/hub4j/github-api/issues/1658 for other complaints related to it.

**Comment by bitwiseman** [comment]
@gsmet 
All the references I'm seeing say [removing checked exceptions doesn't break binary compatibility](https://wiki.eclipse.org/Evolving_Java-based_APIs_2#:~:text=(1)%20Adding%20and%20deleting%20checked,(and%20source%20code%20compatibility).).

**Reply by gsmet** [reply]
Yeah, what's concerning though is that you're breaking the contract if people actually only caught the `IOException`.

Really, it's your decision as the maintainer. I just think the API would be a lot better without having to catch `IOException` everywhere but I can also understand why you're reluctant to it.

My personal inclination would be to do it in 2.0 so that the API gets better with this version. I see a lot of traction for custom automation lately and I wouldn't be surprised if the API had a lot of new users using the library directly instead of hidden behind a Jenkins plugin.

**Reply by MarkEWaite** [reply]
> Really, it's your decision as the maintainer. I just think the API would be a lot better without having to catch `IOException` everywhere but I can also understand why you're reluctant to it.

I think it is a mistake to sacrifice or significantly disrupt the existing users of the library in order to make the API better for users that are not yet consuming the API.  I guess that if the disruption happens and it is too great, then the downstream users (like the Jenkins plugin) can fork the repository and maintain their own fork, but that will reduce the consumers of this API.

**Reply by bitwiseman** [reply]
@gsmet @Haarolean @MarkEWaite 
(Moving to this thread and expanding)
Here's a first cut at a PR removing checked exceptions from public APIs. https://github.com/hub4j/github-api/pull/1969

What the library changes and test changes in this PR indicate to me is that removing checked exceptions may not be a binary breaking change but it is a massive behavior breaking change. Existing consumers of this library depend on exceptions and their types. This change will cause lots of pain. 

Meanwhile, there are several ways to implement the stream-friendly behavior additively, without changing the existing API. If you look at #1658, I mentioned at [one way](https://github.com/hub4j/github-api/issues/1658#issuecomment-1806598640).  If adding `Unckecked` nested classes isn't sufficient, you could create a `org.kohsuke.github.streaming` package that implements stream-friendly wrappers for various classes.

There are also definitely methods where removing the checked exceptions makes sense, but what #1969 makes clear to me is that blanket removal of checked exceptions from the existing API has unacceptably high risk and cost for existing users especially when there are low risk and medium-to-low cost alternatives.

**Reply by gsmet** [reply]
From @MarkEWaite :
> I think it is a mistake to sacrifice or significantly disrupt the existing users of the library in order to make the API better for users that are not yet consuming the API. I guess that if the disruption happens and it is too great, then the downstream users (like the Jenkins plugin) can fork the repository and maintain their own fork, but that will reduce the consumers of this API.

I'm not sure I understand what you mean. This change will improve the life of the existing users: we had many complaints over time and as an extensive user (and contributor) of the API for quite a while, I can't count how many times it has been in the way of some stream API calls or just very annoying because having to deal with `IOException` everywhere.

At every talk I do about GitHub Apps and GitHub Actions, during my live coding, I have to throw a `IOException` for something as simple as adding a comment to an issue. And every time, I say "sorry, that's how it is".

>  then the downstream users (like the Jenkins plugin) can fork the repository and maintain their own fork, but that will reduce the consumers of this API.

They can continue to use the version they are using if they are happy with it. Or they can move to the new version and benefit from the better DevEx.

From Liam:
> Meanwhile, there are several ways to implement the stream-friendly behavior additively, without changing the existing API. If you look at https://github.com/hub4j/github-api/issues/1658, I mentioned at https://github.com/hub4j/github-api/issues/1658#issuecomment-1806598640. If adding Unckecked nested classes isn't sufficient, you could create a org.kohsuke.github.streaming package that implements stream-friendly wrappers for various classes.

So we are fixing something bad in the API by adding wrappers that will make things even more complex? Not exactly sure it's the way to go.

BTW, I'm not saying throwing checked exceptions never makes sense but there are I think very few cases for which it makes sense.
When a label doesn't exist, does it make sense to throw an exception?

This method is a good example of having to handle simple logic with exception control flow: https://github.com/quarkusio/conversational-release-action/blob/main/src/main/java/io/quarkus/bot/release/step/CreateBranch.java#L111 .
I'm not sure it actually helps making things readable and knowing what is a normal error case (this thing doesn't exist) and what is not (I was unable to contact GitHub). It's definitely not fluid when writing code (and having to refer to the Javadoc for every call you write is definitely not ideal).

Especially since I would expect an exception only if for instance I was unable to contact GitHub, not if the thing doesn't exist. The fact that a label doesn't exist is a perfectly valid case from a Java API POV. It shouldn't be an error.

Now if you think it's better to keep the current situation, I can live with it but let's make sure we also acknowledge that we will continue to make things harder than they need to be for the users.
My understanding from the discussions we had was that it would at least be on the table for 2.0, thus why I asked about it.

**Reply by MarkEWaite** [reply]
> From @MarkEWaite :
> 
> > I think it is a mistake to sacrifice or significantly disrupt the existing users of the library in order to make the API better for users that are not yet consuming the API. I guess that if the disruption happens and it is too great, then the downstream users (like the Jenkins plugin) can fork the repository and maintain their own fork, but that will reduce the consumers of this API.
> 
> I'm not sure I understand what you mean. This change will improve the life of the existing users

I'd extend your phrase "This change will improve the life of the existing users" to say "This change will improve the life of the existing users if and only if they rework the logic in their code to adapt to the changes".  In other words, existing users only benefit if they rework the logic in their code.  That's what I meant when I said "sacrifice or significantly disrupt the existing users of the library".

> > then the downstream users (like the Jenkins plugin) can fork the repository and maintain their own fork, but that will reduce the consumers of this API.
> 
> They can continue to use the version they are using if they are happy with it. Or they can move to the new version and benefit from the better DevEx.

Being locked on a specific outdated version has been a bad experience in the Jenkins project.  It creates a situation that looks like a fork of the upstream repository without the clarity of actually forking the upstream repository and releasing it as a new artifact.

When a security issue is detected in the upstream library, the consumer of the specific outdated version must either perform all the work to update to the fixed version (adapting to the behavior change while also performing a security update) or they must fork the repository and backport the security fix to a new release that is based on the specific outdated version.

**Reply by gsmet** [reply]
I understand your point but what you're saying is basically "never do a new major version because changes to the API will be too disruptive for the existing ecosystem".

I'm not saying the situation is ideal, it's definitely not. This change is quite nasty. But it's a bit of a shame that even for a **major** version, we can't do breaking changes when something is actually bad in the API.
Note that I have quite a lot of code using the GitHub API so I know it's going to have an impact on me too. I'm not exactly excited about it but I think this API would be a lot easier to use in the future if we did this change.

I agree it's going to be painful for a while but it might be worth it if it makes the life of our future selves and all the new users better.

> When a security issue is detected in the upstream library, the consumer of the specific outdated version must either perform all the work to update to the fixed version (adapting to the behavior change while also performing a security update) or they must fork the repository and backport the security fix to a new release that is based on the specific outdated version.

That's why you usually push security fixes to multiple release streams - at least when they are not too disruptive.

Again, I won't be the one making the decision and I understand your point, I'm just trying to challenge things a bit because well a major version doesn't happen very often and they are usually quite unique occasions to improve things significantly.

**Reply by Haarolean** [reply]
I agree with Guillaume, if we can't afford to introduce breaking changes even in semver release versions, then, when can we?

**Reply by gsmet** [reply]
BTW, if we decide to go with it, I think we probably need to review very carefully the typology of the changes, categorize them and make sure the semantic is good and consistent for all of them.

I'm preparing Devoxx and will be off all next week but happy to discuss it further after Devoxx.

**Reply by bitwiseman** [reply]
## Major version != free pass for all breaking changes

> I understand your point but what you're saying is basically "never do a new major version because changes to the API will be too disruptive for the existing ecosystem".

No, what's being said is that just because a particular breaking API change has advantages doesn't mean it must be included in a major version bump.  

> I agree with Guillaume, if we can't afford to introduce breaking changes even in semver release versions, then, when can we?

Not all major versions are huge changes. The main reason I'm pulling the trigger on 2.x is to reduce ongoing maintenance headaches: Java 8, `URLConnection`,  and ancient bridge methods.  A major version bump is is not justification in itself for any particular change, it only means more changes can be considered.  

## Streaming-friendly API is an important feature, now how to implement it

I have not heard anyone speak against having a Streaming-friendly API.  The only thing I've said is that it can be implemented without breaking existing contracts.  Yes, that will add code to the project, but from my perspective that code has low complexity. I'm even comfortable saying we should move to have streaming-friendly code being the default for new features.  We progressively deprecate the old code and move the project in that direction so that in a year or so we can cut a 3.0 that drops the old style.   

## Not all API problems related to exceptions are the exceptions fault

> This method is a good example of having to handle simple logic with exception control flow: https://github.com/quarkusio/conversational-release-action/blob/main/src/main/java/io/quarkus/bot/release/step/CreateBranch.java#L111 . I'm not sure it actually helps making things readable and knowing what is a normal error case (this thing doesn't exist) and what is not (I was unable to contact GitHub). It's definitely not fluid when writing code (and having to refer to the Javadoc for every call you write is definitely not ideal).
> 
> Especially since I would expect an exception only if for instance I was unable to contact GitHub, not if the thing doesn't exist. The fact that a label doesn't exist is a perfectly valid case from a Java API POV. It shouldn't be an error.

Ah! This is particularly interesting case.   Yes, this **should** be an exception - you asked for an instance of a specific branch (or label, or whatever) and none was returned.  The API doesn't know that you want to say is, "If branch A doesn't exist create it" or "If  label A exists, rename it.  What we should be talking about is not removing exceptions, but making a more expressive API overall.   For instance, anything that has a `get*` or a `list*` should automatically have an `exists*` method.  But the way the API is written currently makes that a pain to implement.  What if your code could look like this:

```java
if (!repository.branch().name(releaseInformation.getBranch()).exists()) {
    try {
        // If get() fails, we want an exception, not a null.
        String sha = repository.branch().name(releaseInformation.getOriginBranch()).get().getSHA1();
        repository.ref().create()
            .name("refs/heads/" + releaseInformation.getBranch())
            .sha(sha)
            .done();
    } catch (Exception e2) {
        throw new IllegalStateException(
                "Unable to create branch " + releaseInformation.getBranch() + " from "
                        + releaseInformation.getOriginBranch() + ": " + e2.getMessage(),
                e2);
    }
}
```
Rather than getting stuck on modifying the existing methods, let's talk about creating an API that is better.

**Comment by gsmet** [comment]
> No, what's being said is that just because a particular breaking API change has advantages doesn't mean it must be included in a major version bump.

No, sorry, that's not what is being said by some people in the discussion and I was specifically referring to that (sorry, no nested threading creates confusion).
The same "it will break existing users" and "we can't stay on an old version either" arguments made will be there for version 3 or 4 if we don't have a clear plan on how to get there.
We are talking about a library that could be there for the next 20 years. It will have to evolve and there will be major breaking changes.

> Not all major versions are huge changes. The main reason I'm pulling the trigger on 2.x is to reduce ongoing maintenance headaches: Java 8, URLConnection, and ancient bridge methods. A major version bump is is not justification in itself for any particular change, it only means more changes can be considered.

That's exactly what we are asking here. That we consider this change for the new major version.
And that we discuss how it could make the API better in the various cases there is.

For me there are two different discussions:
- what is the API we want if we didn't have the history?
- how do we get there (if there's a way to get there)? Answer might be we use another `groupId`. Answer might be we don't.

If we don't have an honest discussion about the first point, I'm seriously in doubt we will be able to significantly improve ever.

Now if your goal is to release version 2 quickly and that you are not closed to the idea of discussing another major release soonish, that makes sense. But if the idea is that we make a new major every 5 years, then we know that we are stuck for another 5 years.

Our understanding was that, given we were preparing a new major, we could take a deep breath and discuss where we want to go but that might have been our mistake if it wasn't the goal of this version 2.

As the maintainer, it's really your call what you want to do with this library. But as contributors, it's important for us to know what the general plan is to see where we put our efforts.

> Ah! This is particularly interesting case. Yes, this should be an exception - you asked for an instance of a specific branch (or label, or whatever) and none was returned. The API doesn't know that you want to say is, "If branch A doesn't exist create it" or "If label A exists, rename it. What we should be talking about is not removing exceptions, but making a more expressive API overall. For instance, anything that has a get* or a list* should automatically have an exists* method. But the way the API is written currently makes that a pain to implement. What if your code could look like this:

I really don't see your proposal as an improvement API wise but I suppose it's a matter of taste. You are adding a lot of verbosity/specificity to something that is solvable with standard Java constructs.

Also if I want to do something like `getting the branch or creating it if doesn't exist then do something`, I will have one more request in the former case, which is not ideal when being rate-limited. Or you will end up with an `Optional` basically.

BTW, this is exactly what I was asking for earlier, that we try to categorize cases and see how we would like to solve them. Then we can discuss how breaking it is and how we want to get there.
We could also have a look at how other GitHub APIs do it.

There are cases where we know we just want to drop the `IOException` (see the `getHtmlUrl()` case I pointed out), others where we probably want to but that are less obvious and then some for which we will need to think about what we actually want to express.

**Comment by bitwiseman** [comment]
@gsmet 
This has been a long thread with strong opinions expressed by all parties.  

> That's exactly what we are asking here. That we consider this change for the new major version. And that we discuss how it could make the API better in the various cases there is.

You seem to think that I'm not giving your proposals their proper consideration.  I've already devoted a chunk of time to creating a prototype for the "remove all checked exceptions" that folks have suggested.  Yes, my determination from that was to decline that particular proposal for this version, but I put in the good-faith effort to evaluate it.  I am listening to your requests and I'm putting the work and I am open to other proposed changes. 

> Now if your goal is to release version 2 quickly and that you are not closed to the idea of discussing another major release soonish, that 

Yes, my goal (perhaps not clearly stated enough in the descript of this release) is to release version 2 quickly.  I am complete open to making another major release in 6-18 months.  

I would be very comfortable agreeing to a road map where:
* a new API is implemented in a separate package in this library
* elements of the old (current) API are deprecated as the functionality is implemented in the new API
* 6-12 months after the new API achieves feature parity with current API, we cut a major version removing the old API

> ...
> For me there are two different discussions:
> 
> * what is the API we want if we didn't have the history?
> * how do we get there (if there's a way to get there)? Answer might be we use another `groupId`. Answer might be we don't.

So, what are your proposals for these two topics? Especially the first one. 
What would you like https://github.com/quarkusio/conversational-release-action/blob/main/src/main/java/io/quarkus/bot/release/step/CreateBranch.java#L111 to look like ideally? 

> BTW, this is exactly what I was asking for earlier, that we try to categorize cases and see how we would like to solve them. Then we can discuss how breaking it is and how we want to get there. We could also have a look at how other GitHub APIs do it.
> 
> There are cases where we know we just want to drop the `IOException` (see the `getHtmlUrl()` case I pointed out), others where we probably want to but that are less obvious and then some for which we will need to think about what we actually want to express.

Please create an issue to get us started (or add to the existing one, though it's a bit of a long with discussion already). I don't have the bandwidth to lead this charge.

**Reply by gsmet** [reply]
Hey Liam,

Sorry, things have been a bit crazy on my side.

I completely agree with you, I think most of the misunderstanding was about the goals of this version 2.

I'll work on a proposal as time permits, discuss it with my colleagues and try to draw what could be done for the future.

Cheers!

**Comment by bh-dremio** [comment]
Just wanted to send a big thank you for the effort in cleaning up the repo.

The removal of the bridge function annotations saved me a major headache with mockito unit testing.

**Comment by alecharp** [comment]
Thank you @bitwiseman for your effort on this. I'm sorry I haven't had time to try the alpha version so far. I'm catching up.
So far, migration from 1.318 to 2.0.0-alpha-2 is fine. Test migrations to a bit more brain consuming with listing pull requests as we have to deal with the `PageIterable`.

**Reply by alecharp** [reply]
I have to admit that I'm hitting a problem with the `PageIterable`.
I would like to be able to say that for any `GHRepository#queryPullRequests()#list()`, I return some content. I could precise the `toList()` or `toArray()` method, but for some implementation I could prefer to use the `splitIterator()` for a `Stream` implementation.

The problem is that the `PageIterable` implementations are usable, and implementing one is problematic as the `PageIterator` constructor is package visible only, so I cannot use it in a `PageIterable#_iterator()` implementation.

I don't have a solution right now or an idea of a solution but I'll try to think of one.

**Reply by bitwiseman** [reply]
@alecharp 
I've thought about exposing page iterator as public.  But I've been looking to coordinate it with other improvements to page iterable. See #1730 and related issues and discussion. 

Can you give an example of what you'd like the API to look like?

**Reply by alecharp** [reply]
I tent to like the Stream API. 
If we could have a `.stream()`, so that we can easily manipulate the feedback of the API. It could provide an equivalent of the `.toList()` or `toArray()` of the `PageIterator`. 
It could simplify the mocking of such request, no?
